pkg/operator/controller/extension/virtualcluster/reconciler.go

```diff
@@ -0,0 +1,313 @@
+	GardenNamespace string
+}
+
+// Reconcile performs the main reconciliation logic.
+func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log := logf.FromContext(ctx)
+
+	extension := &operatorv1alpha1.Extension{}
+	if err := r.RuntimeClient.Get(ctx, request.NamespacedName, extension); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.V(1).Info("Object is gone, stop reconciling")
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, fmt.Errorf("error retrieving object from store: %w", err)
+	}
+
+	gardenList := &operatorv1alpha1.GardenList{}
+	// We limit one result because we expect only a single Garden object to be there.
+	if err := r.RuntimeClient.List(ctx, gardenList, client.Limit(1)); err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving garden object: %w", err)
+	}
+	if len(gardenList.Items) == 0 {
+		return reconcile.Result{}, nil
+	}
+
+	garden := &gardenList.Items[0]
+	if care.NewGardenConditions(r.Clock, garden.Status).ConvertToMap()[operatorv1alpha1.VirtualGardenAPIServerAvailable].Status != gardencorev1beta1.ConditionTrue {
+		log.Info("virtual garden is not ready yet. Requeue...")
+		return reconcile.Result{}, &reconciler.RequeueAfterError{
+			RequeueAfter: requeueUnhealthyVirtualKubeApiserver,
+		}
+	}
+
+	gardenClientSet, err := r.GardenClientMap.GetClient(ctx, keys.ForGarden(garden))
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("error retrieving garden client object: %w", err)
+	}
+
+	var updatedConditions VirtualClusterConditions
+	if extension.DeletionTimestamp != nil {
+		updatedConditions, err = r.delete(ctx, log, gardenClientSet.Client(), extension)
+	} else {
+		updatedConditions, err = r.reconcile(ctx, log, gardenClientSet.Client(), extension)
+	}
+
+	errUpdate := r.updateExtensionStatus(ctx, log, extension, updatedConditions)
+	if err != nil || errUpdate != nil {
+		return reconcile.Result{}, errors.Join(err, errUpdate)
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *Reconciler) updateExtensionStatus(ctx context.Context, log logr.Logger, extension *operatorv1alpha1.Extension, updatedConditions VirtualClusterConditions) error {
+	currentConditions := NewVirtualClusterConditions(r.Clock, extension.Status)
+	if extension.Generation == extension.Status.ObservedGeneration && !v1beta1helper.ConditionsNeedUpdate(currentConditions.ConvertToSlice(), updatedConditions.ConvertToSlice()) {
+		return nil
+	}
+
+	patch := client.MergeFrom(extension.DeepCopy())
+	// Rebuild garden currentConditions to ensure that only the currentConditions with the correct types will be updated, and any other
+	// currentConditions will remain intact
+	extension.Status.Conditions = v1beta1helper.BuildConditions(extension.Status.Conditions, updatedConditions.ConvertToSlice(), currentConditions.ConditionTypes())
+	extension.Status.ObservedGeneration = extension.Generation
+
+	log.Info("Updating extension status currentConditions")
```

Comment:
```suggestion
	log.V(1).Info("Updating extension status currentConditions", )
```

Please move down, right before patching happens.